### PR TITLE
op-chain-ops: Canyon checker script

### LINKED ISSUE #7731: Testing Plan for Canyon Network Upgrade
The features currently planned for canyon are the following:

* Activating Shapella
* Changing EIP-1559 parameters
* Adjusting the channel ordering in the channel bank
* Adding span batches
* Fixing an op-geth receipt hashing bug
* Adding the create2Deployer on base

We have unit tests for each of these features are part of our normal development process, but would like to test the activation & assert that the behavior is as expected before and after the network upgrade.

To test the above, we want to have scripts which do the following:

* Shapella Before
    * Block RLP **does not** contains withdrawals
    * PUSH0 instruction **does not** work
* Shalella After
    * Block RLP **does** contains withdrawals
    * PUSH0 instruction **does** work
    * P2P gossip continues to work
* EIP-1559 Changes
    * Check that the new parameters are being used only after activation by calculating the basefee by the old formula & also by the new formula 
* Channel Ordering Fix
    * Submit a dangling channel prior to the upgrade & observe a safe head stall
    * Submit a dangling channel after the upgrade & observe no safe head stalls
* Span Batches
    * Submit span batches prior to activation & assert that no L2 blocks are derived from them.
    * Submit span batches after activation & assert that the L2 safe chain advances.
    * Submit normal batches after activation & assert that those still work.
* op-geth receipt hashing bug
    * Manually RLP encode & hash a pre & post canyon deposit receipt. Verify that the hash is in the receipts trie
* create2Deployer
    * Assert that the codehash pre-canyon is the empty code hash. Assert that post-canyon on specific chains that it is the expected code hash
    * If testing the activation of multiple chains, it should remain the empty code hash on some chains
    * Post-Canyon, send a transaction that uses this contract

### PR DESCRIPTION
**Pull request by trianglesphere**
**Description**

This tests the following features of Canyon
1. The receipt hash change
2. The 1559 Parameter change (note: if the basefee is too low this check may not properly report)
3. Check for withdrawals (to indicate Shapella has activated)
4. Check for the create2Deployer code hash

One change to `eth.BlockInfo` to expose the gas limit. TBH I could actually remove it b/c I also dial `ethclient` in addition to `sources.L1Client`.

**Tests**

Manually tested against the devnet.

**Metadata**

- Fixes #7731

### CODE REVIEW

File: op-chain-ops/cmd/check-canyon/main.go [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+	// Define and parse the command-line flags
+	flag.BoolVar(&canyonActive, "canyon", false, "Set this flag to assert canyon behavior")
+	flag.Uint64Var(&number, "number", 31, "Block number to check")
+	flag.Uint64Var(&elasticity, "elasticity", 6, "Specify the EIP-1559 elasticity. 6 on mainnet/sepolia. 10 on goerli")
+	flag.StringVar(&rpcURL, "rpc-url", "http://localhost:8545", "Specify the L2 ETH RPC URL")
+
+	// Parse the command-line arguments
+	flag.Parse()
+
+	l1RPC, err := client.NewRPC(context.Background(), logger, rpcURL, client.WithDialBackoff(10))
```
Comment (ajsutton) [inline op-chain-ops/cmd/check-canyon/main.go]: Is the rpcURL for L1 or L2?

File: op-chain-ops/cmd/check-canyon/main.go [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+}
+
+func ValidateWithdrawals(ctx Args, canyonActive bool) error {
+	block, err := ctx.Client.BlockByNumber(context.Background(), new(big.Int).SetUint64(ctx.Number))
+	if err != nil {
+		return err
+	}
+
+	if canyonActive && block.Withdrawals() == nil {
+		return errors.New("No withdrawals in a canyon block")
```
Comment (ajsutton) [inline op-chain-ops/cmd/check-canyon/main.go]: We could be more strict here and require withdrawals to be the empty array (assuming we're requesting L2 blocks - the client is a bit ambiguous about what it's pointing to).

**Review by ajsutton — APPROVED** [review]
LGTM, just worth clarifying the l1 vs l2 naming.

File: op-chain-ops/cmd/check-canyon/main.go (hunk 1 above)
Comment (trianglesphere) [reply op-chain-ops/cmd/check-canyon/main.go]: L2 url (but treating it more similar to the L1). Will fix

**Pull request merged**